app/services/actions/prospective_student/registration.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module Actions
+  module ProspectiveStudent
+    class Registration
+      Error = Class.new(StandardError) do
+        attr_accessor :details
+      end
+
+      attr_reader :prospective
+
+      def initialize(prospective)
+        @prospective = prospective
+      end
+
+      class << self
+        def call(prospective)
+          new(prospective).call
+        end
+      end
+
+      def call
+        valid!
+
+        ActiveRecord::Base.transaction do
+          user.save!
+          student.save!
+          register!
+          archive!
+          upload_avatar!
+        end
+
+        true
+      end
+
+      private
+
+      delegate :unit,
+               :can_permanently_register?,
+               :id_number,
+               :email,
+               :obs_status,
+               :academic_term,
+               :student_entrance_type,
+               :avatar,
+               to: :prospective
+
+      def valid!
+        return true if user.valid? && student.valid?
+
+        error         = Error.new 'RecordInvalid'
+        error.details = [user.errors.full_messages, student.errors.full_messages].flatten!
+
+        raise error
+      end
+
+      def register!
+        prospective.update!(registered: true)
+      end
+
+      def archive!
+        prospective.update!(archived: true) if user.activated
+      end
+
+      def upload_avatar!
+        return if avatar.blank?
+
+        file = Tempfile.new(['avatar', '.jpg'])
```

Comment:
Standart kitaplıkta [önerilen pratiği](https://ruby-doc.org/stdlib-2.5.3/libdoc/tempfile/rdoc/Tempfile.html#class-Tempfile-label-Explicit+close) izleyelim:

```ruby
begin
  file = Tempfile ...
  ...
ensure
  file.close
  file.unlink
end
```

Reply:
`file.close` dediğim zaman `closed stream` hatası alıyorum o yüzden sadece file.unlink kullanacağım. ActiveStroage dosyayı upload ediyor ama yineden dosya içeriğine ihtiyaç duyuyor sanırsam.

Reply:
Aslında bu tempfile'lar ruby prosesi sonlandığında GC'lenir (unlink); fakat anladığım kadarıyla 7/24 çalışan bir proses olduğundan yığılmayı önlemek için unlink yapıyorlar.  close tarafına ayrıca bakacağım, şimdilik böyle gidelim.